[DataPipe] Extend FileLister to support load multiple directories

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#72260 [DataPipe] Extend FileLister to support load multiple directories**

Change `FileLister` to support single directory, multiple directories and files. This is needed for TorchAudio revamping especially for `OnDiskCacheHolder`

Differential Revision: [D33979744](https://our.internmc.facebook.com/intern/diff/D33979744)

ejguan committed:
[DataPipe] Extend FileLister to support load multiple directories

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

ejguan committed:
Update on "[DataPipe] Extend FileLister to support load multiple directories"


Change `FileLister` to support single directory, multiple directories and files. This is needed for TorchAudio revamping especially for `OnDiskCacheHolder`

Differential Revision: [D33979744](https://our.internmc.facebook.com/intern/diff/D33979744)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

ejguan commented on torch/utils/data/datapipes/iter/__init__.py:
```diff
@@ -1,3 +1,6 @@
+from torch.utils.data.datapipes.iter.utils import (
+    IterableWrapperIterDataPipe as IterableWrapper,
+)
```
Re-order the import to prevent circular import.

NivekT commented on torch/utils/data/datapipes/iter/filelister.py:
```diff
@@ -1,22 +1,24 @@
-        root: Root directory
+        root: Root directory or a sequence of root directories
         mask: Unix style filter string or string list for filtering file name(s)
         abspath: Whether to return relative pathname or absolute pathname
```
Can you update the `args` description to include the arguments `recursive`, `abspath`, `non_deterministic`? Maybe length too?

NivekT replied:
`non_deterministic: if False, the results yielded from each root directory will be sorted`

NivekT replied:
Update `mask` to `masks`

NivekT reviewed (approved):
LGTM with some note. I don't think this is BC-breaking so no need for extra label. Can we add this to 1.11 or does it need to wait?

ejguan commented:
> LGTM with some note. I don't think this is BC-breaking so no need for extra label. Can we add this to 1.11 or does it need to wait?

We may not be able to bring it into 1.11 because branch cut has done and this PR doesn't meet any Cherry-Pick Criteria IMO. We can still let users to use nightly.

ejguan committed:
Update on "[DataPipe] Extend FileLister to support load multiple directories"


Change `FileLister` to support single directory, multiple directories and files. This is needed for TorchAudio revamping especially for `OnDiskCacheHolder`

Differential Revision: [D33979744](https://our.internmc.facebook.com/intern/diff/D33979744)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33979744).

ejguan commented:
@NivekT Second thought. I will try to convince release team to make this PR merged into 1.11. Otherwise, we have to create another `FileLister` in TorchData to override it.
